src/main/java/org/terasology/launcher/gui/javafx/ApplicationController.java

```diff
@@ -440,8 +441,13 @@ private void initButtons() {
         cancelDownloadButton.setTooltip(new Tooltip(BundleUtils.getLabel("launcher_cancelDownload")));
         cancelDownloadButton.managedProperty().bind(cancelDownloadButton.visibleProperty());
         cancelDownloadButton.setVisible(false);
+        
+        if (!selectedPackage.isInstalled()) {
+            startAndDownloadButton.setTooltip(new Tooltip(BundleUtils.getLabel("launcher_download")));
+        } else {
+            startAndDownloadButton.setTooltip(new Tooltip(BundleUtils.getLabel("launcher_start")));
```

Comment:
right, the fact whether this is a download or play action is determined by whether the selected package is installed or not. JavaFX has concept of _bindings_ that allows to bind a value to something else, e.g., the property of being installed. 
I'll have to look up how this actually works in detail, though.

Reply:
So, I explored this path a bit more, and would like to share what I've figured out with you here 🙃 

Ideally, we don't want to care too much about all the places that may influence the state of the button, but rather would like to _derive the tool tip_ from that state. We can use JavaFX bindings on properties to express that:

```java
Property<Tooltip> playButtonTooltip = new SimpleObjectProperty<>(new Tooltip(BundleUtils.getLabel("launcher_download")));
startAndDownloadButton.tooltipProperty().bind(playButtonTooltip);
```

We'll only need this expression once, and afterwards the tool tip will update whenever the `playButtonTooltip` property changes.

But how/when does it change? The one we've created above just holds a static value and never changes.

Reply:
Let's assume for a moment that we have a `selectedPackageProperty` that updates its value whenever the selected package change in the launcher:

```java
private Property<Package> selectedPackageProperty;
```

We can _listen_ to changes to this property by adding a `ChangeListener`. The function given as argument will be called whenever the property changes (somewhat like callbacks in JS):

```java
selectedPackageProperty.addListener((observable, oldValue, newValue) -> {
    // ...
});
```

Inside this callback, we can check the `newValue` (which is a `Package`) whether it is installed or not. Depending on that check, we set the value on the tooltip property:

```java
selectedPackageProperty.addListener((observable, oldValue, newValue) -> {
        Tooltip t = new Tooltip(BundleUtils.getLabel(
                newValue.isInstalled() ?"launcher_start" : "launcher_download"));
        playButtonTooltip.setValue(t);
});
```

Reply:
We assumed that the `selectedPackageProperty` changes whenever the user selects a different game, but that is not the case right now. So, we need to hook that up somewhere. Luckily, searching for _write access_ to `selectedPackage` reveils that it is only changed in a single place inside `initComboBoxes` method.

For the tooltip to work correctly, it is enough to set the value of the property in line 365, too:

```java
selectedPackageProperty.setValue(selectedPackage);
```